File: apps/server2/src/container/container.py
from dependency_injector import containers, providers
from src.api.components.health_check.health_check_service import HealthCheckService
from src.api.components.user.user_repository import UserRepository
from src.api.components.user.user_service import UserService
from src.services.api_pagination_service import APIPaginationService
from src.services.db_service import DBService


class Container(containers.DeclarativeContainer):
    db_service_provider = providers.Singleton(DBService)
    user_repository_provider = providers.Singleton(
        UserRepository, db_service=db_service_provider
    )
    health_check_service_provider = providers.Singleton(
        HealthCheckService, db_service=db_service_provider
    )
    user_service_provider = providers.Singleton(
        UserService, user_repository=user_repository_provider
    )
    api_pagination_service_provider = providers.Singleton(APIPaginationService)

File: apps/server1/src/container/container.ts
import { container, DependencyContainer } from 'tsyringe'

import {
  HealthCheckService,
  IHealthCheckService,
} from '../api/components/health-check'
import {
  IUserRepository,
  IUserService,
  UserRepository,
  UserService,
} from '../api/components/user'
import {
  APIPaginationService,
  IAPIPaginationService,
} from '../services/api-pagination-service'
import { DBService, IDBService } from '../services/db-service'

class Container {
  private readonly _container: DependencyContainer

  constructor() {
    this._container = container
    this._container.register<IDBService>('DBService', {
      useValue: new DBService(),
    })
    this._container.register<IUserRepository>('UserRepository', {
      useValue: new UserRepository(
        this._container.resolve<DBService>('DBService'),
      ),
    })
    this._container.register<IHealthCheckService>('HealthCheckService', {
      useValue: new HealthCheckService(
        this._container.resolve<DBService>('DBService'),
      ),
    })
    this._container.register<IUserService>('UserService', {
      useValue: new UserService(
        this._container.resolve<UserRepository>('UserRepository'),
      ),
    })
    this._container.register<IAPIPaginationService>('APIPaginationService', {
      useValue: new APIPaginationService(),
    })
  }

  public get container(): DependencyContainer {
    return this._container
  }
}

export { Container }
